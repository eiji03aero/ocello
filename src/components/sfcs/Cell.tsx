import * as React from "react";
import styled from "styled-components";

const CellStyled = styled.div`
  background-color: green;
  display: inline-block;
  width: 80px;
  height: 80px;
  margin-bottom: 10px;
  margin-right: 10px;
`;

interface Props {
  children: React.ReactNode;
}

export const Cell: React.SFC = (props: Props) => {
  return (
    <CellStyled>
      { props.children }
    </CellStyled>
  );
};
