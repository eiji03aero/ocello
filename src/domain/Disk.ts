export enum DiskColors {
  Black = "Black",
  White = "White",
}

export interface DiskArgs {
  color: DiskColors;
}

export class Disk {
  color: DiskColors;

  constructor (args: DiskArgs) {
    this.color = args.color;
  }

  static Black () {
    return new Disk({ color: DiskColors.Black });
  }

  static White () {
    return new Disk({ color: DiskColors.White });
  }
}
