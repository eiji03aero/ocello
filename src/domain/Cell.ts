import { Disk, DiskColors } from "./Disk";

export enum CellState {
  Blank = "Blank",
  Placed = "Placed",
}

interface CellArgs {
  state: CellState;
  disk?: Disk;
}

export class Cell {
  state: CellState;
  disk: Disk | null;

  constructor (args: CellArgs) {
    this.state = args.state || CellState.Blank;
    this.disk = args.disk || null;
  }

  static Blank () {
    return new Cell({ state: CellState.Blank });
  }

  static Black () {
    return new Cell({ state: CellState.Placed, disk: Disk.Black() });
  }

  static White () {
    return new Cell({ state: CellState.Placed, disk: Disk.White() });
  }

  static fromColor (color: DiskColors) {
    return new Cell({ state: CellState.Placed, disk: new Disk({ color: color })});
  }

  get color () {
    return this.state === CellState.Placed
      ? this.disk.color
      : '';
  }

  get isBlank () {
    return this.state === CellState.Blank;
  }

  get isPlaced () {
    return this.state === CellState.Placed;
  }
}
