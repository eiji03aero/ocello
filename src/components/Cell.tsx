import * as React from "react";
import * as cn from "classnames";
import styled from "styled-components";

import { Disk } from "./Disk";

import { CellCoordinates } from "../domain/Board";
import { Cell as CellClass } from "../domain/Cell";
import { colors } from "../utils/colors";

const CellStyled = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 80px;
  height: 80px;
  background-color: ${colors.green};
  border: 1px solid ${colors.black};

  &.placeable {
    &:hover {
      cursor: pointer;
      opacity: .8;
    }
  }

  &.not-placeable {
    &:hover {
      pointer-events: none;
    }
  }
`;

export interface Props {
  cell: CellClass;
  rowIndex: number;
  columnIndex: number;
  onPlaceDisk: (coords: CellCoordinates) => void;

  placeable: boolean;
}

export const Cell: React.SFC<Props> = ({
  cell,
  rowIndex,
  columnIndex,
  onPlaceDisk,

  placeable,
}: Props) => {
  const classNames = cn(placeable ? 'placeable' : 'not-placeable')
  const onClickHandler = (e: React.MouseEvent) => {
    if (placeable) {
      onPlaceDisk([rowIndex, columnIndex]);
    }
  }

  return (
    <CellStyled className={classNames}
      onClick={onClickHandler}
    >
      <Disk classNames={cell.color} />
    </CellStyled>
  );
};
