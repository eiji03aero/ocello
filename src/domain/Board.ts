import { Cell } from './Cell';

export type BoardCells = Cell[][];

export type DiskCoordinates = [number, number];

export interface BoardArgs {
  data: BoardCells;
}

export class Board {
  data: BoardCells;

  constructor (args: BoardArgs) {
    this.data = args.data;
  }

  static Blank () { return new Board({ data: this.defaultData }); }
  static Random () { return new Board({ data: this.randomData }); }

  private static get defaultData () {
    return new Array(8).fill(0).map(row => {
      return new Array(8).fill(0).map(col => {
        return Cell.Blank();
      });
    });
  }

  private static get randomData () {
    return new Array(8).fill(0).map(row => {
      return new Array(8).fill(0).map(col => {
        const rand = Math.floor(Math.random() * 3);
        return (
          rand === 0 ? Cell.Blank() :
          rand === 1 ? Cell.Black() :
          Cell.White()
        );
      });
    });
  }
}
