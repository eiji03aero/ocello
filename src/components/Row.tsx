import * as React from "react";
import * as _ from "lodash";
import styled from "styled-components";

import { CellContainer } from "../containers/CellContainer"

import { Cell as CellClass } from "../domain/Cell";
import { BoardRow } from "../domain/Board";

const RowStyled = styled.div`
  display: flex;
`;

interface Props {
  row: BoardRow;
  rowIndex: number;
}

export const Row: React.SFC<Props> = ({
  row,
  rowIndex,
}: Props) => {
  return (
    <RowStyled>
      { _.map(row, (cell: CellClass, idx) => (
        <CellContainer
          key={idx}
          cell={cell}
          rowIndex={rowIndex}
          columnIndex={idx}
        />
      ))}
    </RowStyled>
  );
};
