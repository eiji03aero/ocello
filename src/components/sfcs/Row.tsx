import * as React from "react";
import * as _ from "lodash";
import styled from "styled-components";

import { Cell as CellClass } from "../../domain/Cell";
import { Cell } from "./Cell"

const RowStyled = styled.div`
  display: flex;
`;

interface Props {
  cells: CellClass[];
}

export const Row: React.SFC<Props> = ({
  cells,
}: Props) => {
  return (
    <RowStyled>
      { _.map(cells, (cell, idx) => (
        <Cell key={idx} state={cell.state} />
      ))}
    </RowStyled>
  );
};
