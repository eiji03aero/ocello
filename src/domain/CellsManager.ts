import * as _ from "lodash";
import { CellCoordinates, BoardCells } from "./Board";

export class CellManager {
  static LowerEnd = 0;
  static HigherEnd = 7;

  static getAround (coords: CellCoordinates): CellCoordinates[] {
    const seedCoords = [
      [-1, -1], [0, -1], [1, -1],
      [-1, 0], [1, 0],
      [-1, 1], [0, 1], [1, 1],
    ];
    const candidateCoords = _.map(seedCoords, (sc: CellCoordinates) => {
      const rc = coords[0] + sc[0];
      const cc = coords[1] + sc[1];
      return [rc, cc] as CellCoordinates;
    });
    return _.filter(candidateCoords, (cand: CellCoordinates) => {
      return (
        CellManager.LowerEnd <= cand[0] &&
        CellManager.LowerEnd <= cand[1] &&
        CellManager.HigherEnd >= cand[0] &&
        CellManager.HigherEnd >= cand[1]
      );
    });
  }

  static checkIfPlaceable (coords: CellCoordinates, data: BoardCells) {
    return true
  }
}
