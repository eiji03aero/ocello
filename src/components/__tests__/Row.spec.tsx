import * as React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as _ from "lodash";

import { Row } from "../Row";
import { Cell } from "../../domain/Cell";

describe('Row', () => {
  it('matches snapshot', () => {
    const row = _.map(_.fill(new Array(8), 0), (_: number) => Cell.Blank());
    const wrapper = shallow(
      <Row
        row={row}
        rowIndex={0}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
