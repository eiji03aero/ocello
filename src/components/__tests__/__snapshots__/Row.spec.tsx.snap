// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row matches snapshot 1`] = `
<styled.div>
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={0}
    key="0"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={1}
    key="1"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={2}
    key="2"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={3}
    key="3"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={4}
    key="4"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={5}
    key="5"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={6}
    key="6"
    rowIndex={0}
  />
  <CellContainer
    cell={
      Cell {
        "disk": null,
        "state": "Blank",
      }
    }
    columnIndex={7}
    key="7"
    rowIndex={0}
  />
</styled.div>
`;
